### PR TITLE
Refactor Editor Container 

### PR DESCRIPTION
**Pull request by AdamMeza-Bentley**
## Changes

This PR makes a number of changes to the `EditorContainer` component. 
- Migrates away from class based to functional React syntax as [outlined in the contributing docs](https://github.com/iTwin/appui/blob/master/CONTRIBUTING.md#common-issues-to-solve). 

## Testing

Component functionality has been tested in standalone and storybook. 

## Notes
This PR began by addressing a bug outlined in - Addresses the issues outline in #630.  However, after refactoring the component and attempting the solution provided it still displayed some odd behavior with the escape key event. This bug will be solved in a future PR

### CODE REVIEW

**Comment by AdamMeza-Bentley** [comment]
I noticed the use of `React.cloneElement`. The [React Docs](https://react.dev/reference/react/cloneElement#alternatives) strong urge against using it. I'm looking into the other ways they give. Are there any prefernces? @GerardasB. 

Screenshot 2024-03-28 101055

**Comment by AdamMeza-Bentley** [comment]
> > I noticed the use of `React.cloneElement`. The [React Docs](https://react.dev/reference/react/cloneElement#alternatives) strongly urge against using it. I'm looking into the other ways they give. Are there any prefernces?
> 
> Not sure if we can do much w/o breaking existing editors. I'd leave it as a separate task for later.

Ok sweet! I can make a separate issue

File: ui/components-react/src/components-react/editors/EditorContainer.tsx [hunk 1]
```diff
@@ -130,78 +145,104 @@ export class EditorContainer extends React.PureComponent<EditorContainerProps> {
...
 
     return clonedNode;
-  }
+  };
 
-  private _handleEditorBlur = (_e: React.FocusEvent) => {
+  /** Event Handlers
+   * @internal
+   */
+  const handleCancel = () => onCancel();
```
Comment (GerardasB) [inline ui/components-react/src/components-react/editors/EditorContainer.tsx:153]: Remove alias, use `onCancel` instead.

File: ui/components-react/src/components-react/editors/EditorContainer.tsx [hunk 2]
```diff
@@ -218,91 +259,43 @@ export class EditorContainer extends React.PureComponent<EditorContainerProps> {
...
+      className="components-editor-container"
+      onBlur={handleContainerBlur}
+      onKeyDown={handleKeyDown}
+      onClick={handleClick}
+      onContextMenu={handleClick}
+      title={title}
+      data-testid="editor-container"
+      role="presentation"
+    >
+      {createEditor()}
```
Comment (GerardasB) [inline ui/components-react/src/components-react/editors/EditorContainer.tsx:298]: Remove this function call, inline it in the component body instead.

File: ui/components-react/src/components-react/editors/EditorContainer.tsx [hunk 3]
```diff
@@ -130,78 +145,104 @@ export class EditorContainer extends React.PureComponent<EditorContainerProps> {
...
+        newValue,
+      });
+    }
   };
 
-  private _handleContainerBlur = (e: React.FocusEvent) => {
-    e.stopPropagation();
+  const onPressEscape = (): void => {
+    // istanbul ignore else
+    if (propertyEditor && propertyEditor.containerHandlesEscape) handleCancel();
```
Comment (GerardasB) [inline ui/components-react/src/components-react/editors/EditorContainer.tsx:172]: This seems to match the previous handler.
I can't seem to reproduce the #630 here https://itwin.github.io/appui/storybook/?path=/docs/components-editors--docs
What was the cause and what changed to fix it?

**Review by GerardasB — APPROVED** [review]

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 2 above)
Comment (AdamMeza-Bentley) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:298]: DO you mean inline the whole editor creation? that seem excessive as its 30+ lines.

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 3 above)
Comment (AdamMeza-Bentley) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:172]: I'm not sure to be honest. I wasn't able to recreate it either but I made the switch that was recommended

File: ui/components-react/src/components-react/editors/EditorContainer.tsx [hunk 4]
```diff
@@ -91,37 +91,52 @@ export interface TypeEditor {
...
+    setFocus,
+    onCommit,
+    onCancel,
+  } = props;
+
+  let propertyEditor: PropertyEditorBase;
+
+  const editorRef = React.useRef<TypeEditor | null>(null);
+
+  const setEditorRef = (ref: TypeEditor | null) => (editorRef.current = ref);
```
Comment (ignas-k) [inline ui/components-react/src/components-react/editors/EditorContainer.tsx:111]: Seems like a good place to use `React.useState`

File: ui/components-react/src/components-react/editors/EditorContainer.tsx [hunk 5]
```diff
@@ -130,78 +145,104 @@ export class EditorContainer extends React.PureComponent<EditorContainerProps> {
...
-  private _handleContainerBlur = (e: React.FocusEvent) => {
-    e.stopPropagation();
+  const onPressEscape = (): void => {
+    // istanbul ignore else
+    if (propertyEditor && propertyEditor.containerHandlesEscape) handleCancel();
+  };
+
+  const onPressEnter = (e: React.KeyboardEvent): void => {
+    // istanbul ignore else
+    if (propertyEditor && propertyEditor.containerHandlesEnter) {
```
Comment (ignas-k) [inline ui/components-react/src/components-react/editors/EditorContainer.tsx:177]: Maybe we can write `propertyEditor?. propertyEditor.containerHandlesEnter` for a cleaner solution? You should make this change to all similar places

**Review by ignas-k — APPROVED** [review]

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 5 above)
Comment (AdamMeza-Bentley) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:177]: yeah thats a good catch. i'll change it in the conditionals that use that syntax

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 2 above)
Comment (GerardasB) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:291]: It's 30lines + function declaration (that's used once) either way.

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 3 above)
Comment (GerardasB) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx]: Should we update the PR description then if we are not sure that it fixes the #630?
Were you able to reproduce that issue before your changes?

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 2 above)
Comment (AdamMeza-Bentley) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:291]: thas a good point! I just wanted to double check before I made that change

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 4 above)
Comment (ignas-k) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:110]: I am actually curious about this myself. Could you try `const [editorRef, setEditorRef] = React.useState(React.useRef<TypeEditor | null>(null));` and see if it works fine?

Comment (GerardasB) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:110]: Please don't, what's wrong with the ref?

Comment (ignas-k) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx:110]: Nothing wrong, just thought that we could combine the two things into one

File: ui/components-react/src/components-react/editors/EditorContainer.tsx (hunk 3 above)
Comment (AdamMeza-Bentley) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx]: I changed the PR description to address that. I did try the suggested change but it wasn't successful

File: ui/components-react/src/components-react/editors/EditorContainer.tsx [hunk 6]
```diff
@@ -218,91 +216,77 @@ export class EditorContainer extends React.PureComponent<EditorContainerProps> {
...
+      className="components-editor-container"
+      onBlur={handleContainerBlur}
+      onKeyDown={handleKeyDown}
+      onClick={handleClick}
+      onContextMenu={handleClick}
+      title={title}
+      data-testid="editor-container"
+      role="presentation"
+    >
+      {() => {
```
[the remark this answers is not in the document]
Comment (AdamMeza-Bentley) [reply ui/components-react/src/components-react/editors/EditorContainer.tsx]: fixed this.

**Pull request merged**